File: Wells/Source/Filters/FilterCollection.vb
Imports System.Collections.ObjectModel

Public Class FilterCollection(Of T)
    Inherits ObservableCollection(Of BaseFilter(Of T))

    Protected _PropertyName As String

    Event FiltersChanged()

    Sub RaiseCollectionChanged()
        RaiseEvent FiltersChanged()
    End Sub

    Function Apply(list As IQueryable(Of T)) As IQueryable(Of T)
        For Each f In Me
            If f.IsEnabled Then
                list = f.Apply(list)
            End If
        Next
        Return list
    End Function
End Class

File: DesktopApp/src/Filters/FilterCollection.cs
using System;
using System.Collections.ObjectModel;
using System.Linq;


namespace Wells.View.Filters
{
    public class FilterCollection<T> : ObservableCollection<BaseFilter<T>>
    {

        protected string PropertyName;

        public EventHandler FiltersChanged { get; set; }
        public void RaiseCollectionChanged()
        {
            FiltersChanged(this, EventArgs.Empty);
        }

        public IQueryable<T> Apply(IQueryable<T> queryable)
        {
            foreach (var f in this)
            {
                if (f.IsEnabled)
                    queryable = f.Apply(queryable);
            }
            return queryable;
        }
    }
}
